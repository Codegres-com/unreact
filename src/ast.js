// Root.
export const rootName = 'Program';
export const createRoot = () => ({
  type: rootName,
  children: []
});

// HTML elements.
export const elementName = 'Element';
export const createElement = (tagName, attributes = []) => ({
  type: elementName,
  tagName,
  attributes,
  children: []
});

// HTML text elements.
export const textName = 'Text';
export const createText = value => ({
  type: textName,
  value
});

// HTML attributes.
export const attributeName = 'Attribute';
export const createAttribute = (name, value, expression = false) => ({
  type: attributeName,
  name,
  value,
  expression
});

// Template engine syntax for inline JavaScript.
export const scriptletName = 'Scriptlet';
export const createScriptlet = value => ({
  type: scriptletName,
  value
});

// Template engine syntax for escaped interpolation.
export const interpolationEscapedName = 'InterpolationEscaped';
export const createInterpolationEscaped = value => ({
  type: interpolationEscapedName,
  value
});

// Template engine syntax for unescaped interpolation.
export const interpolationUnescapedName = 'InterpolationUnescaped';
export const createInterpolationUnescaped = value => ({
  type: interpolationUnescapedName,
  value
});

// Template engine syntax for condition.
export const conditionName = 'Condition';
export const createCondition = (test, consequent, alternate = null) => ({
  type: conditionName,
  test,
  consequent,
  alternate
});

// Template engine syntax for iteration.
export const iterationName = 'Iteration';
export const createIteration = (test, consequent, alternate = null) => ({
  type: iterationName,
  test,
  consequent,
  alternate
});
