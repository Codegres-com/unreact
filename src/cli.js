/* eslint-disable no-console */
import yargs from 'yargs';
import chalk from 'chalk';
import ora from 'ora';
import { compileFile, compileDir } from './index';
import pkg from '../package.json';

export default function unreactCLI(argv) {
  console.log(chalk.bold.white(`${pkg.name} v${pkg.version}`));

  const parsedArgv = yargs(argv)
    .option('o', {
      alias: 'out-file',
      describe: 'Output file',
      type: 'string'
    })
    .option('O', {
      alias: 'out-dir',
      describe: 'Output dir',
      type: 'string'
    })
    .option('t', {
      alias: 'template-engine',
      describe: 'Output template engine',
      type: 'string',
      default: 'pug',
      choices: ['pug', 'ejs']
    })
    .option('add-beginning', {
      describe: 'Add string to the beginning of the output file',
      type: 'string'
    })
    .option('add-ending', {
      describe: 'Add string to the ending of the output file',
      type: 'string'
    })
    .option('initial-indent-level', {
      describe: 'Start at <number> indent level',
      type: 'number',
      default: 0
    })
    .usage(`${pkg.description}.\nUsage: $0 <file or dir> [options]`)
    .version()
    .alias('version', 'v')
    .help()
    .alias('help', 'h').argv;
  const fileOrDir = parsedArgv._[0];
  const spinner = ora({ text: 'Processing...' });
  const progress = report => {
    spinner.stopAndPersist({ text: `${chalk.gray(report)}` });
  };
  const templateEngine = parsedArgv.t;
  const beginning = parsedArgv['add-beginning'];
  const ending = parsedArgv['add-ending'];
  const initialIndentLevel = parsedArgv['initial-indent-level'];

  if (fileOrDir && parsedArgv.o) {
    spinner.start();
    return compileFile(fileOrDir, parsedArgv.o, {
      templateEngine,
      beginning,
      ending,
      initialIndentLevel,
      progress
    })
      .then(() => {
        spinner.succeed(`${chalk.bold.green('success')} file transformed to ${templateEngine}`);
      })
      .catch(err => {
        spinner.stopAndPersist();
        spinner.fail(`${chalk.bold.red('error')} ${err.stack}`);
        process.exit(1);
      });
  }

  if (fileOrDir && parsedArgv.O) {
    spinner.start();
    return compileDir(fileOrDir, parsedArgv.O, {
      templateEngine,
      beginning,
      ending,
      initialIndentLevel,
      progress
    })
      .then(() => {
        spinner.succeed(`${chalk.bold.green('success')} folder transformed to ${templateEngine}`);
      })
      .catch(err => {
        spinner.stopAndPersist();
        spinner.fail(`${chalk.bold.red('error')} ${err.message}`);
        process.exit(1);
      });
  }

  return null;
}
