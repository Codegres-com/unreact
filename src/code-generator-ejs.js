/* eslint-disable no-param-reassign */
import htmlTagsVoids from 'html-tags/void';
import {
  attributeName,
  conditionName,
  elementName,
  interpolationEscapedName,
  iterationName,
  mixinName,
  rootName,
  textName
} from './ast';

function codeGeneratorEjs(node, { initialIndentLevel = 0, indentLevel = initialIndentLevel } = {}) {
  switch (node.type) {
    case rootName:
      return node.children
        .map(child => codeGeneratorEjs(child, { initialIndentLevel, indentLevel }))
        .join('');
    case mixinName:
      return node.children.map(child =>
        codeGeneratorEjs(child, { initialIndentLevel, indentLevel })
      );
    case elementName:
      return indent(
        generateTag(
          node.tagName,
          node.children
            .map(child =>
              codeGeneratorEjs(child, { initialIndentLevel, indentLevel: indentLevel + 1 })
            )
            .join(''),
          node.attributes
            .map(child =>
              codeGeneratorEjs(child, { initialIndentLevel, indentLevel: indentLevel + 1 })
            )
            .join('')
        ),
        {
          initialIndentLevel,
          indentLevel
        }
      );
    case textName:
      return node.value;
    case attributeName:
      return generateProperty(node.name, node.value, node.expression);
    case interpolationEscapedName:
      return generateInterpolationEscaped(node.value);
    case conditionName:
      return indent(
        generateCondition(
          node.test,
          codeGeneratorEjs(node.consequent, { initialIndentLevel, indentLevel: indentLevel + 1 }),
          node.alternate &&
            codeGeneratorEjs(node.alternate, { initialIndentLevel, indentLevel: indentLevel + 1 })
        ),
        {
          initialIndentLevel,
          indentLevel
        }
      );
    case iterationName:
      return indent(
        generateIteration({
          iterable: node.iterable,
          currentValue: node.currentValue,
          index: node.index,
          array: node.array,
          body: codeGeneratorEjs(node.body, { initialIndentLevel, indentLevel: indentLevel + 1 })
        }),
        {
          initialIndentLevel,
          indentLevel
        }
      );
    default:
      throw new TypeError(node.type);
  }
}

export default codeGeneratorEjs;

function generateTag(tagName, children, properties) {
  const startTagBeginning = `<${tagName}${properties}`;
  if (htmlTagsVoids.includes(tagName)) {
    return `${startTagBeginning} />`;
  }
  const startTag = `${startTagBeginning}>`;
  const endTag = `</${tagName}>`;
  const tag = startTag + children + endTag;

  return tag;
}

function generateProperty(name, value, expression) {
  const normalizedName = normalizePropertyName(name);
  const startPropertyBeginning = ` ${normalizedName}`;

  // NOTE: `value === true` is to accept boolean attributes, e.g.: `<input checked />`.
  if (value === true) {
    return startPropertyBeginning;
  }

  if (expression) {
    return `${generateScriptlet(
      `if (![null,undefined].includes(${value})) {`
    )}${startPropertyBeginning}="${generateInterpolationEscaped(value)}"${generateScriptlet('}')}`;
  }

  return `${startPropertyBeginning}="${value}"`;
}

function normalizePropertyName(name) {
  switch (name) {
    case 'className':
      return 'class';
    case 'htmlFor':
      return 'for';
    case 'tabIndex':
      return 'tabindex';
    default:
      return name;
  }
}

function generateCondition(test, consequent, alternate) {
  const conditionArray = [
    generateScriptlet(`if (${test}) {`),
    consequent,
    alternate ? generateScriptlet('} else {') : null,
    alternate,
    generateScriptlet('}')
  ].filter(Boolean);
  return conditionArray.join('');
}

function generateIteration({ iterable, currentValue, index, array, body }) {
  const params = [currentValue, index, array].filter(Boolean).join(', ');
  const iterationArray = [
    generateScriptlet(`${iterable}.forEach((${params}) => {`),
    body,
    generateScriptlet('})')
  ].filter(Boolean);
  return iterationArray.join('');
}

function generateScriptlet(value) {
  return `<% ${value} %>`;
}

function generateInterpolationEscaped(value) {
  return `<%= ${value} %>`;
}

function indent(str, { initialIndentLevel, indentLevel }) {
  const indentChar = ' ';
  const indentLength = 2;
  const startIndentNumber = indentLevel * indentLength;
  const endIndentNumber = (indentLevel ? indentLevel - 1 : indentLevel) * indentLength;
  const strIndented = `${indentLevel === initialIndentLevel ? '' : '\n'}${indentChar.repeat(
    startIndentNumber
  )}${str}${'\n'}${indentChar.repeat(endIndentNumber)}`;
  return strIndented;
}
