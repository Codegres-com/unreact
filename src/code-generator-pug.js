/* eslint-disable no-param-reassign */
import htmlTagsVoids from 'html-tags/void';
import {
  attributeName,
  conditionName,
  elementName,
  interpolationEscapedName,
  iterationName,
  mixinName,
  rootName,
  textName
} from './ast';

function codeGeneratorPug(
  node,
  { initialIndentLevel = 0, indentLevel = initialIndentLevel, previousSibling = null } = {}
) {
  switch (node.type) {
    case rootName:
      return node.children
        .map(child => codeGeneratorPug(child, { initialIndentLevel, indentLevel }))
        .join('');
    case mixinName:
      return node.children.map(child =>
        codeGeneratorPug(child, { initialIndentLevel, indentLevel })
      );
    case elementName:
      return indent(
        generateTag(
          node.tagName,
          node.children
            .map((child, i) =>
              codeGeneratorPug(child, {
                initialIndentLevel,
                indentLevel: indentLevel + 1,
                previousSibling: i > 0 ? node.children[i - 1] : null
              })
            )
            .join(''),
          node.attributes
            .map(child =>
              codeGeneratorPug(child, { initialIndentLevel, indentLevel: indentLevel + 1 })
            )
            .join(', ')
        ),
        {
          initialIndentLevel,
          indentLevel
        }
      );
    case textName:
      if (previousSibling && previousSibling.type === elementName) {
        return indent(`| ${node.value}`, { initialIndentLevel, indentLevel });
      }
      return node.value;
    case attributeName:
      return generateProperty(node.name, node.value, node.expression);
    case interpolationEscapedName:
      if (previousSibling && previousSibling.type === elementName) {
        return indent(`| ${generateInterpolationEscaped(node.value)}`, {
          initialIndentLevel,
          indentLevel
        });
      }
      return generateInterpolationEscaped(node.value);
    case conditionName:
      return indent(
        generateCondition(
          node.test,
          codeGeneratorPug(node.consequent, { initialIndentLevel, indentLevel: indentLevel + 1 }),
          node.alternate &&
            codeGeneratorPug(node.alternate, { initialIndentLevel, indentLevel: indentLevel + 1 }),
          initialIndentLevel,
          indentLevel + 1
        ),
        {
          initialIndentLevel,
          indentLevel
        }
      );
    case iterationName:
      return indent(
        generateIteration({
          iterable: node.iterable,
          currentValue: node.currentValue,
          index: node.index,
          array: node.array,
          body: codeGeneratorPug(node.body, { initialIndentLevel, indentLevel: indentLevel + 1 }),
          initialIndentLevel,
          indentLevel
        }),
        {
          initialIndentLevel,
          indentLevel
        }
      );
    default:
      throw new TypeError(node.type);
  }
}

export default codeGeneratorPug;

function generateTag(tagName, children, properties) {
  const startTag = `${tagName}${properties ? '(' : ''}${properties}${properties ? ')' : ''}`;
  if (htmlTagsVoids.includes(tagName)) {
    return startTag;
  }
  const addSpace = children && !children[0].includes('\n');
  const tag = startTag + (addSpace ? ` ${children}` : children);
  return tag;
}

function generateProperty(name, value, expression) {
  const normalizedName = normalizePropertyName(name);

  // NOTE: `value === true` is to accept boolean attributes, e.g.: `<input checked />`.
  if (value === true) {
    return normalizedName;
  }

  if (expression) {
    return `${normalizedName}=${value}`;
  }

  return `${normalizedName}="${value}"`;
}

function normalizePropertyName(name) {
  switch (name) {
    case 'className':
      return 'class';
    case 'htmlFor':
      return 'for';
    case 'tabIndex':
      return 'tabindex';
    default:
      return name;
  }
}

function generateCondition(test, consequent, alternate, initialIndentLevel, indentLevel) {
  const newConsequent =
    consequent[0] === '\n' ? consequent : indent(consequent, { initialIndentLevel, indentLevel });
  const alternateOrNull = stuff => (alternate ? stuff() : null);
  const conditionArray = [
    `if ${test}`,
    newConsequent,
    alternateOrNull(() => indent('else', { initialIndentLevel, indentLevel: indentLevel - 1 })),
    alternateOrNull(
      () =>
        alternate[0] === '\n' ? alternate : indent(alternate, { initialIndentLevel, indentLevel })
    )
  ].filter(Boolean);
  return conditionArray.join('');
}

function generateIteration({
  iterable,
  currentValue,
  index,
  array,
  body,
  initialIndentLevel,
  indentLevel
}) {
  const params = [currentValue, index].filter(Boolean).join(', ');
  const iterationArray = [
    `each ${params} in ${iterable}`,
    array
      ? indent(generateScriptlet(`const ${array} = ${iterable};`), {
          initialIndentLevel,
          indentLevel: indentLevel + 1
        })
      : null,
    body
  ].filter(Boolean);
  return iterationArray.join('');
}

function generateScriptlet(value) {
  return `- ${value}`;
}

function generateInterpolationEscaped(value) {
  return `#{${value}}`;
}

function indent(str, { initialIndentLevel, indentLevel }) {
  const indentChar = ' ';
  const indentLength = 2;
  const startIndentNumber = indentLevel * indentLength;
  const isRoot = indentLevel === initialIndentLevel;
  const strIndented = `${isRoot ? '' : '\n'}${indentChar.repeat(startIndentNumber)}${str}${
    isRoot ? '\n' : ''
  }`;
  return strIndented;
}
