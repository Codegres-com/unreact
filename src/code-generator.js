import isSelfClosing from 'is-self-closing';
import {
  elementName,
  rootName,
  textName,
  attributeName,
  interpolationEscapedName,
  conditionName,
  iterationName
} from './ast';

function codeGenerator(node, level = 0, removeEmptyLine = false) {
  switch (node.type) {
    case rootName:
      return node.children.map((child, i) => codeGenerator(child, level, i === 0)).join('');
    case elementName:
      return indent(
        generateTag(
          node.tagName,
          node.children.map(child => codeGenerator(child, level + 1, removeEmptyLine)).join(''),
          node.attributes.map(child => codeGenerator(child, level + 1, removeEmptyLine)).join('')
        ),
        level,
        removeEmptyLine
      );
    case textName:
      return node.value;
    case attributeName:
      return generateProperty(node.name, node.value, node.expression);
    case interpolationEscapedName:
      return generateInterpolationEscaped(node.value);
    case conditionName:
      return indent(
        generateCondition(
          node.test,
          codeGenerator(node.consequent, level + 1, removeEmptyLine),
          node.alternate && codeGenerator(node.alternate, level + 1, removeEmptyLine)
        ),
        level,
        removeEmptyLine
      );
    case iterationName:
      return indent(
        generateIteration({
          iterable: node.iterable,
          currentValue: node.currentValue,
          index: node.index,
          array: node.array,
          body: codeGenerator(node.body, level + 1, removeEmptyLine)
        }),
        level,
        removeEmptyLine
      );
    default:
      throw new TypeError(node.type);
  }
}

export default codeGenerator;

function generateTag(tagName, children, properties) {
  const startTagBeginning = `<${tagName}${properties}`;
  if (isSelfClosing(tagName)) {
    return `${startTagBeginning} />`;
  }
  const startTag = `${startTagBeginning}>`;
  const endTag = `</${tagName}>`;
  const tag = startTag + children + endTag;

  return tag;
}

function generateProperty(name, value, expression) {
  const normalizedName = normalizePropertyName(name);
  const startPropertyBeginning = ` ${normalizedName}`;

  // NOTE: `value === true` is to accept boolean attributes, e.g.: `<input checked />`.
  if (value === true) {
    return startPropertyBeginning;
  }

  if (expression) {
    return `${startPropertyBeginning}="${generateInterpolationEscaped(value)}"`;
  }

  return `${startPropertyBeginning}="${value}"`;
}

function normalizePropertyName(name) {
  switch (name) {
    case 'className':
      return 'class';
    case 'htmlFor':
      return 'for';
    case 'tabIndex':
      return 'tabindex';
    default:
      return name;
  }
}

function generateCondition(test, consequent, alternate) {
  const conditionArray = [
    generateScriptlet(`if (${test}) {`),
    consequent,
    alternate ? generateScriptlet('} else {') : null,
    alternate,
    generateScriptlet('}')
  ].filter(Boolean);
  return conditionArray.join('');
}

function generateIteration({ iterable, currentValue, index, array, body }) {
  const params = [currentValue, index, array].filter(Boolean).join(', ');
  const iterationArray = [
    generateScriptlet(`${iterable}.forEach((${params}) => {`),
    body,
    generateScriptlet('})')
  ].filter(Boolean);
  return iterationArray.join('');
}

function generateScriptlet(value) {
  return `<% ${value} %>`;
}

function generateInterpolationEscaped(value) {
  return `<%= ${value} %>`;
}

function indent(str, level, removeEmptyLine) {
  const indentChar = ' ';
  const indentLength = 2;
  const startIndentNumber = level * indentLength;
  const endIndentNumber = (level ? level - 1 : level) * indentLength;
  const strIndented = `${removeEmptyLine && level === 0 ? '' : '\n'}${indentChar.repeat(
    startIndentNumber
  )}${str}${'\n'}${indentChar.repeat(endIndentNumber)}`;
  return strIndented;
}
