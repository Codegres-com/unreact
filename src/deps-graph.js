export default class DependencyGraph {
  constructor() {
    this.modules = [];
  }

  addModule(path, ast, table) {
    return this.modules.push({
      path,
      ast,
      table,
      dependencies: []
    });
  }

  find(path) {
    return this.modules.find(node => node.path === path);
  }

  addDependency(startPath, endPath) {
    const startNode = this.find(startPath);
    const endNode = this.find(endPath);

    if (!startNode || !endNode) {
      throw new Error('Both modules need to exist');
    }

    startNode.dependencies.push(endNode);
  }
}
