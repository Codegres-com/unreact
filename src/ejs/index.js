import codeGenerator from './code-generator';
import optimize from './optimize';

export { optimize, codeGenerator };
