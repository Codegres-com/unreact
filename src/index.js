import fs from 'fs';
import path from 'path';
import util from 'util';
import makeDir from 'make-dir';
import globby from 'globby';
import parser from './parser';
import transformation from './transformation';
import { codeGenerator, optimize } from './ejs';

export function compile(inputCode) {
  const oldAst = parser(inputCode);
  const { ast: newAST, table } = transformation(oldAst);
  const optimizedAST = optimize(newAST, table);
  const code = codeGenerator(optimizedAST);
  return code;
}

const readFileAsync = util.promisify(fs.readFile);
const writeFileAsync = util.promisify(fs.writeFile);

export async function compileFile(inputFile, outputFile, progress = () => {}) {
  const inputFilePath = path.resolve(process.cwd(), inputFile);
  const outputFilePath = path.resolve(process.cwd(), outputFile);
  const inputCode = await readFileAsync(inputFilePath, { encoding: 'utf8' });
  const code = compile(inputCode);
  const outputFilePathDir = path.dirname(outputFilePath);
  await makeDir(outputFilePathDir);
  await writeFileAsync(outputFilePath, code);
  const relativePath = path.join(process.cwd(), '/');
  progress(`${inputFile.replace(relativePath, '')} -> ${outputFile.replace(relativePath, '')}`);
}

export async function compileDir(inputDir, outputDir, progress = () => {}) {
  const inputDirPath = path.resolve(process.cwd(), inputDir);
  const outputDirPath = path.resolve(process.cwd(), outputDir);
  const globInputDirPath = path.join(inputDirPath, '/**/*.{js,jsx,mjs}');
  const filePaths = await globby(globInputDirPath);
  await Promise.all(
    filePaths.map(filePath => {
      const oldExtension = path.extname(filePath);
      const outputFile = filePath
        .replace(inputDirPath, outputDirPath)
        .replace(oldExtension, '.ejs');
      return compileFile(filePath, outputFile, progress);
    })
  );
}
