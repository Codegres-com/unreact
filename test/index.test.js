import fs from 'fs';
import path from 'path';
import util from 'util';
import { compile, compileDir, compileFile } from '../src/index';

describe('while using react2ejs.compile()', () => {
  it('should convert react components of one tag to ejs', async () => {
    const { input, output } = await getTestCase('one-tag');
    const result = compile(input);
    expect(result).toBe(output);
  });
  it('should convert react components of multiple nested tags to ejs', async () => {
    const { input, output } = await getTestCase('nested-tags');
    const result = compile(input);
    expect(result).toBe(output);
  });
  it('should convert react components of one tag with text to ejs', async () => {
    const { input, output } = await getTestCase('with-text');
    const result = compile(input);
    expect(result).toBe(output);
  });
  it('should convert react components of one tag with var expression to ejs', async () => {
    const { input, output } = await getTestCase('with-var-expression');
    const result = compile(input);
    expect(result).toBe(output);
  });
  it('should convert react components of one tag with complex expression to ejs', async () => {
    const { input, output } = await getTestCase('with-complex-expression');
    const result = compile(input);
    expect(result).toBe(output);
  });
  it('should convert react components of a self closing tag to ejs', async () => {
    const { input, output } = await getTestCase('self-closing-tag');
    const result = compile(input);
    expect(result).toBe(output);
  });
  it('should convert react components with attributes to ejs', async () => {
    const { input, output } = await getTestCase('attrs');
    const result = compile(input);
    expect(result).toBe(output);
  });
  it('should convert react components with boolean attributes to ejs', async () => {
    const { input, output } = await getTestCase('attrs-boolean');
    const result = compile(input);
    expect(result).toBe(output);
  });
  it('should convert react components with var expression attributes to ejs', async () => {
    const { input, output } = await getTestCase('attrs-var-expression');
    const result = compile(input);
    expect(result).toBe(output);
  });
  it('should convert react components with complex expression attributes to ejs', async () => {
    const { input, output } = await getTestCase('attrs-complex-expression');
    const result = compile(input);
    expect(result).toBe(output);
  });
  it('should convert react components with logical expression to ejs', async () => {
    const { input, output } = await getTestCase('logical-expression');
    const result = compile(input);
    expect(result).toBe(output);
  });
  it('should convert react components with conditional expression to ejs', async () => {
    const { input, output } = await getTestCase('conditional-expression');
    const result = compile(input);
    expect(result).toBe(output);
  });
  it('should convert react components with map iterator to ejs', async () => {
    const { input, output } = await getTestCase('iterator');
    const result = compile(input);
    expect(result).toBe(output);
  });
  it('should convert react components to ejs and ignore unnecessary attributes', async () => {
    const { input, output } = await getTestCase('ignore-unnecessary-attrs');
    const result = compile(input);
    expect(result).toBe(output);
  });
  it('should convert react components with logical expression inside conditional expression to ejs', async () => {
    const { input, output } = await getTestCase('logical-expr-inside-conditional-expr');
    const result = compile(input);
    expect(result).toBe(output);
  });
});

describe('while using react2ejs.compileFile()', () => {
  it('should convert react components of one tag to ejs', async () => {
    const fixture = 'compile-file';
    const inputPath = getFixturePath(`${fixture}/input.js`);
    const outputPath = getFixturePath(`${fixture}/output.ejs`);
    await compileFile(inputPath, outputPath);
    const output = await getFixture(`${fixture}/output.ejs`);
    expect(output).toMatchSnapshot();
  });
});

describe('while using react2ejs.compileDir()', () => {
  it('should convert react components of one tag to ejs', async () => {
    const fixture = 'compile-dir';
    const inputOutputPath = getFixturePath(fixture);
    await compileDir(inputOutputPath, inputOutputPath);
    const Nested = await getFixture(`${fixture}/Nested.ejs`);
    const Simple = await getFixture(`${fixture}/Simple.ejs`);
    expect(Nested).toMatchSnapshot();
    expect(Simple).toMatchSnapshot();
  });
});

async function getTestCase(folder) {
  const input = await getFixture(`${folder}/input.js`);
  const output = await getFixture(`${folder}/output.ejs`);
  return { input, output };
}

const readFileAsync = util.promisify(fs.readFile);

async function getFixture(file) {
  const filePath = getFixturePath(file);
  const fileContent = await readFileAsync(filePath, { encoding: 'utf8' });
  return fileContent;
}

function getFixturePath(file) {
  const filePath = path.resolve(__dirname, 'fixtures', file);
  return filePath;
}
