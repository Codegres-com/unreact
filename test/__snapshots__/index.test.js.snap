// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`while using react2ejs.compileDir() should convert react components of one tag to ejs 1`] = `
"<div>
  <div>
    <span></span>
  </div>
</div>
"
`;

exports[`while using react2ejs.compileDir() should convert react components of one tag to ejs 2`] = `
"<div></div>
"
`;

exports[`while using react2ejs.compileFile() should convert react components of one tag to ejs 1`] = `
"<div></div>
"
`;
